File: backend/app/schemas/customers.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal, Optional

from fastapi import UploadFile
from pydantic import BaseModel, ConfigDict, EmailStr, Field

from ..models import KYCStatus


class CustomerRegisterForm(BaseModel):
    """CLAUDE.md §27 (M11): the credit-officer-driven registration path —
    unlike the self-signup ProfileSubmitForm, the full client PRD §2A field
    set is required here, since an officer is capturing a complete in-branch
    intake. Must stay the ONLY body parameter on the route (same
    Form()-flattening constraint as ProfileSubmitForm, see its docstring).
    """

    model_config = ConfigDict(arbitrary_types_allowed=True)

    # User account fields — same shape self-signup already requires (CLAUDE.md
    # §4: role/company_id/branch_id are still always derived server-side,
    # never from this body).
    email: EmailStr
    password: str = Field(min_length=8)
    full_name: str = Field(min_length=1)

    # Identity
    first_name: str = Field(min_length=1)
    middle_name: Optional[str] = None
    last_name: str = Field(min_length=1)
    id_type: Literal["national_id", "passport"] = "national_id"
    national_id_number: str = Field(min_length=1)
    date_of_birth: date
    gender: str = Field(min_length=1)
    nationality: str = Field(min_length=1)
    marital_status: str = Field(min_length=1)
    dependants_count: int = Field(ge=0)

    # Contact
    phone_number: str = Field(min_length=1)
    phone_number_alt: Optional[str] = None
    residential_address: str = Field(min_length=1)

    # Employment
    employment_status: str = Field(min_length=1)
    occupation: str = Field(min_length=1)
    monthly_income: Decimal = Field(gt=0)

    # Next of kin (referees are a separate list — see RefereeInput/the
    # POST /customers/{id}/referees endpoint, added after the customer exists)
    next_of_kin_name: str = Field(min_length=1)
    next_of_kin_relationship: str = Field(min_length=1)
    next_of_kin_phone: str = Field(min_length=1)

    # Documents — same handling as ProfileSubmitForm via app/kyc_storage.py.
    id_document: UploadFile
    id_document_back: UploadFile
    selfie_photo: Optional[UploadFile] = None


class CustomerResponse(BaseModel):
    """Summary shape for the registration response and the customer list —
    composed from User + Profile, so built manually in the router rather
    than via from_attributes on a single ORM object."""

    id: int  # User.id
    profile_id: int
    customer_number: Optional[str]
    full_name: str
    email: str
    branch_id: Optional[int]
    kyc_status: KYCStatus
    created_at: datetime


class RefereeInput(BaseModel):
    full_name: str = Field(min_length=1)
    phone_number: str = Field(min_length=1)
    relationship: Optional[str] = None
    address: Optional[str] = None


class RefereeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    profile_id: int
    full_name: str
    phone_number: str
    relationship: Optional[str]
    address: Optional[str]
    created_at: datetime


class BusinessAssessmentRequest(BaseModel):
    """CLAUDE.md §27: net_income/debt_service_capacity are deliberately NOT
    fields here — app/customer_registration.py::compute_business_figures is
    the only writer of those two, server-side, every time."""

    business_name: str = Field(min_length=1)
    business_type: str = Field(min_length=1)
    ownership: str = Field(min_length=1)
    physical_location: str = Field(min_length=1)
    years_in_operation: int = Field(ge=0)
    sales_frequency: Literal["daily", "weekly", "monthly"]
    total_income: Decimal = Field(ge=0)
    total_expenses: Decimal = Field(ge=0)
    reported_profit: Optional[Decimal] = None
    stock_value: Optional[Decimal] = None
    existing_loans_amount: Optional[Decimal] = None
    other_lenders: Optional[str] = None
    bank_mpesa_turnover: Optional[Decimal] = None
    business_assets_value: Optional[Decimal] = None
    cash_flow_notes: Optional[str] = None
    existing_debt_obligations: Decimal = Field(default=Decimal("0.00"), ge=0)


class BusinessAssessmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    profile_id: int
    business_name: str
    business_type: str
    ownership: str
    physical_location: str
    years_in_operation: int
    sales_frequency: str
    total_income: Decimal
    total_expenses: Decimal
    reported_profit: Optional[Decimal]
    stock_value: Optional[Decimal]
    existing_loans_amount: Optional[Decimal]
    other_lenders: Optional[str]
    bank_mpesa_turnover: Optional[Decimal]
    business_assets_value: Optional[Decimal]
    cash_flow_notes: Optional[str]
    existing_debt_obligations: Decimal
    net_income: Decimal
    debt_service_capacity: Decimal
    created_at: datetime
    updated_at: datetime

File: frontend/src/schemas/customers.ts
import { z } from "zod";

// Fixed choice lists for the customer registration form's categorical
// fields — the backend stores each as a plain string (app/models/profile.py
// has no DB-level enum for these), so this is a frontend-only constraint:
// present a closed set instead of free text, rather than a data-model
// change. Values are the exact strings sent to and stored by the backend.
export const GENDER_OPTIONS = [
  { value: "male", label: "Male" },
  { value: "female", label: "Female" },
] as const;

export const NATIONALITY_OPTIONS = [
  { value: "Kenya", label: "Kenya" },
  { value: "Uganda", label: "Uganda" },
  { value: "Tanzania", label: "Tanzania" },
] as const;

export const MARITAL_STATUS_OPTIONS = [
  { value: "married", label: "Married" },
  { value: "single", label: "Single" },
  { value: "divorced", label: "Divorced" },
] as const;

export const NEXT_OF_KIN_RELATIONSHIP_OPTIONS = [
  { value: "father", label: "Father" },
  { value: "mother", label: "Mother" },
  { value: "brother", label: "Brother" },
  { value: "sister", label: "Sister" },
  { value: "spouse", label: "Spouse" },
  { value: "child", label: "Child" },
] as const;

export const EMPLOYMENT_STATUS_OPTIONS = [
  { value: "employed", label: "Employed" },
  { value: "unemployed", label: "Unemployed" },
  { value: "self_employed", label: "Self Employed" },
] as const;

const MIN_CUSTOMER_AGE_YEARS = 18;

function isAtLeastYearsOld(dateString: string, years: number): boolean {
  const dob = new Date(dateString);
  if (Number.isNaN(dob.getTime())) return false;
  const today = new Date();
  const cutoff = new Date(today.getFullYear() - years, today.getMonth(), today.getDate());
  return dob <= cutoff;
}

// Mirrors backend app/schemas/customers.py
export const customerRegisterSchema = z.object({
  email: z.string().email("Enter a valid email"),
  password: z.string().min(8, "At least 8 characters"),
  full_name: z.string().min(1, "Required"),

  first_name: z.string().min(1, "Required"),
  middle_name: z.string().optional(),
  last_name: z.string().min(1, "Required"),
  id_type: z.enum(["national_id", "passport"]),
  national_id_number: z.string().min(1, "Required"),
  date_of_birth: z
    .string()
    .min(1, "Required")
    .refine((v) => isAtLeastYearsOld(v, MIN_CUSTOMER_AGE_YEARS), {
      message: "Customer must be at least 18 years old",
    }),
  gender: z.enum(["male", "female"], { message: "Required" }),
  nationality: z.enum(["Kenya", "Uganda", "Tanzania"], { message: "Required" }),
  marital_status: z.enum(["married", "single", "divorced"], { message: "Required" }),
  dependants_count: z.string().min(1, "Required"),

  phone_number: z.string().min(1, "Required"),
  phone_number_alt: z.string().optional(),
  residential_address: z.string().min(1, "Required"),

  employment_status: z.enum(["employed", "unemployed", "self_employed"], { message: "Required" }),
  occupation: z.string().min(1, "Required"),
  monthly_income: z.string().min(1, "Required"),

  next_of_kin_name: z.string().min(1, "Required"),
  next_of_kin_relationship: z.enum(["father", "mother", "brother", "sister", "spouse", "child"], {
    message: "Required",
  }),
  next_of_kin_phone: z.string().min(1, "Required"),
});
export type CustomerRegisterInput = z.infer<typeof customerRegisterSchema>;

export interface CustomerResponse {
  id: number;
  profile_id: number;
  customer_number: string | null;
  full_name: string;
  email: string;
  branch_id: number | null;
  kyc_status: "pending" | "verified" | "rejected";
  created_at: string;
}

export const refereeInputSchema = z.object({
  full_name: z.string().min(1, "Required"),
  phone_number: z.string().min(1, "Required"),
  relationship: z.string().optional(),
  address: z.string().optional(),
});
export type RefereeInputForm = z.infer<typeof refereeInputSchema>;

export interface RefereeResponse {
  id: number;
  profile_id: number;
  full_name: string;
  phone_number: string;
  relationship: string | null;
  address: string | null;
  created_at: string;
}

export const businessAssessmentSchema = z.object({
  business_name: z.string().min(1, "Required"),
  business_type: z.string().min(1, "Required"),
  ownership: z.string().min(1, "Required"),
  physical_location: z.string().min(1, "Required"),
  years_in_operation: z.string().min(1, "Required"),
  sales_frequency: z.enum(["daily", "weekly", "monthly"]),
  total_income: z.string().min(1, "Required"),
  total_expenses: z.string().min(1, "Required"),
  reported_profit: z.string().optional(),
  stock_value: z.string().optional(),
  existing_loans_amount: z.string().optional(),
  other_lenders: z.string().optional(),
  bank_mpesa_turnover: z.string().optional(),
  business_assets_value: z.string().optional(),
  cash_flow_notes: z.string().optional(),
  existing_debt_obligations: z.string().optional(),
});
export type BusinessAssessmentInput = z.infer<typeof businessAssessmentSchema>;

export interface BusinessAssessmentResponse {
  id: number;
  profile_id: number;
  business_name: string;
  business_type: string;
  ownership: string;
  physical_location: string;
  years_in_operation: number;
  sales_frequency: string;
  total_income: string;
  total_expenses: string;
  reported_profit: string | null;
  stock_value: string | null;
  existing_loans_amount: string | null;
  other_lenders: string | null;
  bank_mpesa_turnover: string | null;
  business_assets_value: string | null;
  cash_flow_notes: string | null;
  existing_debt_obligations: string;
  net_income: string;
  debt_service_capacity: string;
  created_at: string;
  updated_at: string;
}
